contracts/CharityChallenge.sol

```diff
@@ -89,12 +99,48 @@ contract CharityChallenge {
             isEventFinalizedAndValid = true;
             if (hasChallengeAccomplished) {
                 uint256 totalContractBalance = address(this).balance;
-                npoAddress.transfer(address(this).balance);
-                emit Donated(npoAddress, totalContractBalance);
+                uint8 sumRatio = 0;
+                uint length = npoAddresses.length;
+                for (uint i = 0; i < length; i++) {
+                  address payable npo = npoAddresses[i];
+                  sumRatio += npoRatios[npo];
+                }
+                uint256 donatedAmount = 0;
+                for (uint i = 0; i < length - 1; i++) {
+                  address payable npo = npoAddresses[i];
+                  uint8 ratio = npoRatios[npo];
+                  uint256 amount = totalContractBalance * ratio / sumRatio;
+                  donatedAmount += amount;
+                  npo.transfer(amount);
+                  emit Donated(npo, amount);
+                }
+                // Don't want to keep any amount in the contract
+                uint256 remainingAmount = totalContractBalance - donatedAmount;
+                address payable npo = npoAddresses[length - 1];
+                npo.transfer(remainingAmount);
+                emit Donated(npo, remainingAmount);
             }
         }
     }
 
+    function getExpectedDonationAmount(address payable _npo) view external returns (uint256) {
+      uint256 totalContractBalance = address(this).balance;
+      uint8 sumRatio = 0;
```

Comment:
I think we can add a new instance variable sumRatio and calculate it  in the constructor.
This method can be small ```require(npoRatios[_npo] > 0); return totalContractBalance * npoRatios[_npo] / sumRatio;```